girder/utility/path.py

```diff
@@ -79,3 +89,109 @@ def join(tokens):
+        ('file', parentType == 'item', {'name': token, 'itemId': parent['_id']}),
+    )
+
+    for candidateModel, mask, filterObject in searchTable:
+        if not mask:
+            continue
+
+        candidateChild = ModelImporter.model(candidateModel).findOne(filterObject)
+        if candidateChild is not None:
+            return candidateChild, candidateModel
+
+    # if no folder, item, or file matches, give up
+    raise NotFoundException('Child resource not found: %s(%s)->%s' % (
+        parentType, parent.get('name', parent.get('_id')), token))
+
+
+def lookUpPath(path, user=None, test=False, filter=True):
+    """
+    Look up a resource in the data hierarchy by path.
+
+    :param path: path of the resource
+    :param user: user with correct privileges to access path
+    :param test: defaults to false, when set to true
+        will return None instead of throwing exception when
+        path doesn't exist
+    :type test: bool
+    :param filter: Whether the returned model should be filtered.
+    :type filter: bool
+    """
+    path = path.lstrip('/')
+    pathArray = split(path)
+    model = pathArray[0]
+
+    if model == 'user':
+        username = pathArray[1]
+        parent = ModelImporter.model('user').findOne({'login': username})
+
+        if parent is None:
+            if test:
+                return {
+                    'model': None,
+                    'document': None
+                }
+            else:
+                raise NotFoundException('User not found: %s' % username)
+
+    elif model == 'collection':
+        collectionName = pathArray[1]
+        parent = ModelImporter.model('collection').findOne({'name': collectionName})
+
+        if parent is None:
+            if test:
+                return {
+                    'model': None,
+                    'document': None
+                }
+            else:
+                raise NotFoundException('Collection not found: %s' % collectionName)
+
+    else:
+        raise ValidationException('Invalid path format')
+
+    try:
+        document = parent
+        ModelImporter.model(model).requireAccess(document, user)
+        for token in pathArray[2:]:
+            document, model = lookUpToken(token, model, document)
+            ModelImporter.model(model).requireAccess(document, user)
+    except (ValidationException, AccessException):
+        # We should not distinguish the response between access and validation errors so that
+        # adversarial users cannot discover the existence of data they don't have access to by
+        # looking up a path.
+        if test:
+            return {
+                'model': None,
+                'document': None
+            }
+        else:
+            raise NotFoundException('Path not found: %s' % path)
+
+    if filter:
+        document = ModelImporter.model(model).filter(document, user)
+
+    return {
+        'model': model,
+        'document': document
+    }
```

Comment:
They are covered by existing tests; this method already existed prior to this branch, it was simply moved out from an internal function inside the REST layer into a reusable part of the python API.